### ISSUE TITLE
Logfile overrun by "Failed to query password file entry for user"

### ISSUE THREAD

**Issue by GoogleCodeExporter** [dun/munge#26]
It is the one below in `./src/munged/gids.c`

``` c
log_msg (LOG_INFO,
      "Failed to query password file entry for \"%s\"", user);
return (-1);
```

We use the same group file on every system, but only the subset of the users that needs access to the system appear in passwd, so this particular error causes a lot of output (~3.5MB/hour).

I'm just going to remove it and recompile, but it might be nice to have that error only output with a debug option or there be an option to disable info level logging or something else that seems reasonable to you. Thanks,

Stephen

_Original issue reported on code.google.com by `sjf4@uw.edu` on 23 Oct 2013 at 7:05_

**Comment by sknigh** [comment]
Hi, I am from the future!

Are their any plans for a permanent fix that doesn't require syslog filtering? My cluster has a large group file and curated passwd file. We are currently storing munge logs locally, and it generates about 1 MB of logfile per second per node which caught us off guard.

**Comment by dun** [comment]
So you've been sent back in time to prevent the system from being overrun by rampant logfiles? It's worse than I feared.

Is the logfile continually growing at 1MB/sec? Because this update is only occurring hourly by default.

I still think missing users should be logged because that is generally an unexpected condition; you're operating in a somewhat non-standard configuration.

But munged should ideally only log each missing user the first time it goes missing during the lifetime of the process. And you should also be able to specify the minimum syslog level of messages to log (e.g., LOG_NOTICE and above to ignore informational messages like this). Would those address your situation?

**Comment by sknigh** [comment]
Thanks for the suggestions.

I am using syslog-ng, and I had to add three rules to the default filter to prevent /var/log/messages from getting spammed, due to how broadly it normally captures:
``` bash
filter f_default    { level(info..emerg) and
                        not (facility(mail)
                        or facility(authpriv)
                        or facility(cron))
                        and not match("Unable to query password file entry for")
                        and not match("Unable to query group info")
                        and not match("Failed to query password file entry for"); };
```
The filters could probably be written intelligently to check the sending program first but this will do for now.
 
When I start the munge service (which shouldn't happen often anyway) it will peg a core to about 60% while it finds all the problems with my group file for about 2m, then settles down again.

**Comment by dun** [comment]
The strings "Unable to query password file entry for" and "Unable to query group info" do not occur in version 0.5.11 (released in 2013) or later.  They were changed in 9cc1e047f3f0cad7977de773807118c9feeb69fa.

By default, munged processes the passwd & group files hourly. If you start munged with `--group-update-time=0`, it will only process them once at start. If you set that to `-1`, it will disable this processing altogether (which will prevent you from restricting credentials by GID).

Out of curiosity, how many entries are in your passwd & group files?

**Comment by dun** [comment]
@sknigh: 25acfc95946b6513a0bfd48e6d0fd04747f5410b adds a couple mitigations that should help your situation. Since your setup results in numerous non-existent users, negative caching of the user lookups should significantly reduce the supplementary group processing time. And each non-existent user will only be logged once. I've added notes to Issue #58 for options to suppress log messages for non-existent users and to specify the minimum syslog level.

Also, the log message has changed to:
```C
log_msg (LOG_INFO, "Failed to query passwd file for \"%s\": User not found", user);                 
```
so you might want to filter on "User not found" until the config options appear.

**Comment by sknigh** [comment]
Thanks!

I looked into my current munge installation, and it comes from Centos 7's EPEL repository.

``` bash
$ munge --version
munge-0.5.11 (2013-08-27)
```

``` bash
$ yum info munge
...
Installed Packages
Name        : munge
Arch        : x86_64
Version     : 0.5.11
Release     : 3.el7
Size        : 236 k
Repo        : installed
From repo   : Centos7:epel
Summary     : Enables uid & gid authentication across a host cluster
URL         : http://munge.googlecode.com/
License     : GPLv3+ and LGPLv3+
Description : MUNGE (MUNGE Uid 'N' Gid Emporium) is an authentication service for creating
            : and validating credentials. It is designed to be highly scalable for use
            : in an HPC cluster environment.
            : It allows a process to authenticate the UID and GID of another local or
            : remote process within a group of hosts having common users and groups.
            : These hosts form a security realm that is defined by a shared cryptographic
            : key. Clients within this security realm can create and validate credentials
            : without the use of root privileges, reserved ports, or platform-specific
            : methods.
```

Needless to say it was ancient, which is expected from RHEL/Centos packages.

**Comment by yfoo** [comment]
Hi, we just noticed this effecting us on our HPC cluster. I'm embarrassed that i just noticed this now actually. We see:

...
2019-11-14 11:22:46 Info:      Unable to query password file entry for "acr358"
2019-11-14 11:22:46 Error:     Unable to query group info: Numerical result out of range
...

Hammering the log file on the hour. This creates GB size files. 

Our cluster is setup to look at LDAP for user accounts on the system. The odd thing is that munge is seeing users that are out of scope for those which have access on the system.

I see that you've fixed this issue in 0.5.11, yet we are stuck on 0.5.10 for the moment. We will consider compiling from source, and installing. Yet, is there a way utilizing the 0.5.10 epel rpm to configure munge with possibly /etc/sysconfig/munge to deal with this? Or maybe another config option?

Best,
Chris

**Comment by dun** [comment]
If you want to stay on 0.5.10 for now, you could add `--group-update-time=0` to the munged command-line (i.e., appending that to `DAEMON_ARGS` in `/etc/sysconfig/munge`). That option will cause the processing of the supplementary group membership map to only happen once at startup. You could also set that option to `-1` if you want to disable the supplementary group mapping altogether.

If you want to compile from source, I'd recommend waiting another week or so. I'll be releasing 0.5.14 soon. It includes a couple noteworthy mitigations in 25acfc95946b6513a0bfd48e6d0fd04747f5410b: 1) negative caching to eliminate repeated queries for non-existent users thereby decreasing processing time and reducing log messages, and 2) only logging a message the first time a given non-existent user is detected.
